File: src/main/java/cs/ut/engine/FileWriter.java
package cs.ut.engine;

import cs.ut.config.MasterConfiguration;
import cs.ut.config.nodes.DirectoryPathConfiguration;
import cs.ut.exceptions.NirdizatiRuntimeException;
import org.apache.log4j.Logger;
import org.json.JSONObject;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;

public class FileWriter {
    private static final Logger log = Logger.getLogger(FileWriter.class);

    private String scriptDir;
    private String userModelDir;
    private String coreDir;
    private String datasetDir;
    private String trainingDir;
    private String pklDir;

    public FileWriter() {
        DirectoryPathConfiguration pathProvider = MasterConfiguration.getInstance().getDirectoryPathConfiguration();
        scriptDir = pathProvider.getScriptDirectory();
        userModelDir = pathProvider.getUserModelDirectory();
        coreDir = scriptDir.concat("core/");
        datasetDir = pathProvider.getDatasetDirectory();
        trainingDir = pathProvider.getTrainDirectory();
        pklDir = pathProvider.getPklDirectory();
    }


    public void writeJsonToDisk(JSONObject json, String filename, String path) {
        log.debug("Writing json to disk...");

        File file = new File(coreDir.concat(path.concat(filename.concat(".json"))));

        byte[] bytes;
        try {
            bytes = json.toString().getBytes("UTF-8");
        } catch (UnsupportedEncodingException e) {
            log.debug(String.format("Unsupported encoding %s", e));
            throw new NirdizatiRuntimeException(e);
        }

        if (!file.exists()) {
            try {
                Files.createFile(Paths.get(file.getAbsolutePath()));
                log.debug(String.format("Created file <%s>", file.getName()));
            } catch (IOException e) {
                throw new NirdizatiRuntimeException(String.format("Failed creating file <%s>", file.getAbsolutePath()), e);
            }
        }

        try (OutputStream os = new FileOutputStream(file)) {
            os.write(bytes);
            os.close();
            log.debug(String.format("Successfully written json to disk... <%s> bytes written", bytes.length));
        } catch (IOException e) {
            throw new NirdizatiRuntimeException(String.format("Failed writing json file to disk <%s>", file.getName()), e);
        }
    }
}

File: src/main/kotlin/cs/ut/util/FileWriter.kt
package cs.ut.util

import cs.ut.config.MasterConfiguration
import cs.ut.exceptions.NirdizatiRuntimeException
import org.apache.log4j.Logger
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.charset.Charset
import java.nio.file.Files
import java.nio.file.Paths

class FileWriter {

    private val log = Logger.getLogger(FileWriter::class.java)

    private val pathProvider = MasterConfiguration.directoryPathConfiguration
    private val scriptDir = pathProvider.scriptDirectory
    private val coreDir = scriptDir + "core/"

    fun writeJsonToDisk(json : JSONObject, fileName : String, path : String): File {
        log.debug("Writing json <$json> to disk...")

        val file = File(coreDir + path + fileName + ".json")

        val bytes = json.toString().toByteArray(Charset.forName("UTF-8"))

        if (!file.exists()) {
            try {
                Files.createFile(Paths.get(file.absolutePath))
                log.debug("Successfully create file $file")
            } catch (e : IOException) {
                throw NirdizatiRuntimeException("Failed to create file ${file.absolutePath}")
            }
        }

        FileOutputStream(file).use { it.write(bytes) }
        log.debug("Successfully written json to disk. <${bytes.size}> bytes were written")

        return file
    }
}